google/cloud/aiplatform/_pipeline_based_service/pipeline_based_service.py

```diff
@@ -0,0 +1,328 @@
+    _format_resource_name_method = "pipeline_job_path"
+
+    _valid_done_states = _PIPELINE_COMPLETE_STATES
+
+    @property
+    @classmethod
+    @abc.abstractmethod
+    def _template_ref(self) -> Dict[str, str]:
+        """A dictionary of the pipeline template URLs for this servicewhere the key is
+        an identifier for that template and the value is the url of that pipeline template.
+
+        For example: {"tabular_classification": "gs://path/to/tabular/pipeline/template.json"}
+
+        """
+        pass
+
+    @property
+    @classmethod
+    @abc.abstractmethod
+    def _creation_log_message(self) -> str:
+        """A log message to use when the Pipeline-based Service is created, since this class
+        supresses logs from PipelineJob creation to avoid duplication.
+
+        For example: 'Created PipelineJob for your Model Evaluation.'
+
+        """
+        pass
+
+    # TODO: Consider updating this to return a dict or list in the future to support multiple outputs
+    @property
+    @abc.abstractmethod
+    def _metadata_output_artifact(self) -> Optional[str]:
+        """The ML Metadata output artifact resource URI from the completed pipeline run."""
+        pass
+
+    @property
+    def backing_pipeline_job(self) -> "pipeline_jobs.PipelineJob":
+        """The PipelineJob associated with the resource."""
+        return pipeline_jobs.PipelineJob.get(resource_name=self.resource_name)
+
+    @property
+    def pipeline_console_uri(self) -> Optional[str]:
+        """The console URI of the PipelineJob created by the service."""
+        if self.backing_pipeline_job:
+            return self.backing_pipeline_job._dashboard_uri()
+
+    @property
+    def state(self) -> Optional[gca_pipeline_state.PipelineState]:
+        """The state of the Pipeline run associated with the service."""
+        if self.backing_pipeline_job:
+            return self.backing_pipeline_job.state
+        return None
+
+    # TODO: expose _template_ref in error message when artifact registry support is added
+    def _validate_pipeline_template_matches_service(
+        self, pipeline_job: "pipeline_jobs.PipelineJob"
+    ):
+        """Utility function to validate that the passed in pipeline matches
+        the template of the Pipeline Based Service.
+
+        Args:
+            pipeline_job (aiplatform.PipelineJob):
+                Required. The PipelineJob to validate with this Pipeline Based Service.
+
+        Raises:
+            ValueError: if the provided pipeline ID doesn't match the pipeline service.
+        """
+
+        pipeline_match = False
+
+        for pipeline_template in self._template_ref.values():
+            service_pipeline_json = yaml_utils.load_yaml(pipeline_template)[
+                "components"
+            ]
+            current_pipeline_json = pipeline_job.to_dict()["pipelineSpec"]["components"]
+
+            if service_pipeline_json == current_pipeline_json:
```

Comment:
I've seen `template_uri` empty in a lot of recent PipelineJobs I've submitted. I could add an additional check for it here in cases where `_gca_resource.template_uri` is not None - is that what you meant?

I can also expose it on `PipelineJob` if the Pipelines team plans to use it in the future.

Reply:
Yes, I think a check on the `template_uri` match should be sufficient. But defer to @chensun @IronPan on guidance.

I think that should be the case because AR is not being used yet but it should become more common in the future.

I initially though `template_uri` was only intended to be populated in the AR case but it seems it can be populated with any uri based on this change: https://github.com/googleapis/python-aiplatform/pull/1683/files#diff-61b397e85f75bea63573834d2b088d9404ebf58306b1fbd9f57aee4f5a19e8a2R284

So we can perhaps just use `template_uri` for all cases and not require matching on the PipelineSpec. Defer to @chensun and @IronPan on guidance here as well.

Reply:
Depending on the goal of the comparison. If this is security, core CUJ, or usage tracking related, then a more strict comparison on the content is needed. 
If it's just some non core purpose, template_uri might be ok. 

I might lack the context here in terms of how it's being used, to provide much useful info :)